privval/tcp.go

```diff
@@ -123,6 +123,9 @@ func (sc *TCPVal) OnStart() error {
 						"Ping",
 						"err", err,
 					)
+					sc.conn.Close()
+					sc.RemoteSignerClient = NewRemoteSignerClient(sc.conn)
```

Comment:
If we're closing the `sc.conn`, why do we create a new remote signer client with it? Shouldn't we call OnStop(), then OnStart() to wait for conn and recreate all necessary structures and goroutines?

Shouldn't we assert the err's type? We may want to shut everything down in case of `ErrUnexpectedResponse`, no?

Reply:
> If we're closing the sc.conn, why do we create a new remote signer client with it? Shouldn't we call OnStop(), then OnStart() to wait for conn and recreate all necessary structures and goroutines?

I think that makes sense! 

> Shouldn't we assert the err's type? We may want to shut everything down in case of ErrUnexpectedResponse, no?

Probably yes. `ErrUnexpectedResponse` means the other side responded but with another message than expected (not `PingResponse`).

Reply:
I'll continue here: https://github.com/tendermint/tendermint/pull/2923

Reply:
BTW calling OnStart won't be enough as this might time out (again) if the remote signer isn't up yet. We might need sth like retrying with exponential backoff. 
Also, the whole ConsensusState shuts down if the remote signer drops (restarting the remote signer won't help with this): https://github.com/tendermint/tendermint/issues/2926